kernel/src/checkpoint/mod.rs

```diff
@@ -1,8 +1,509 @@
+use log_replay::CheckpointLogReplayProcessor;
+
+use url::Url;
+
+mod log_replay;
+#[cfg(test)]
+mod tests;
+
+const SECONDS_PER_MINUTE: u64 = 60;
+const MINUTES_PER_HOUR: u64 = 60;
+const HOURS_PER_DAY: u64 = 24;
+const DAYS: u64 = 7;
+/// The default retention period for deleted files in seconds.
+/// This is set to 7 days, which is the default in delta-spark.
+const DEFAULT_RETENTION_SECS: u64 = SECONDS_PER_MINUTE * MINUTES_PER_HOUR * HOURS_PER_DAY * DAYS;
+
+/// Schema for extracting relevant actions from log files for checkpoint creation
+static CHECKPOINT_ACTIONS_SCHEMA: LazyLock<SchemaRef> = LazyLock::new(|| {
+    StructType::new([
+        Option::<Add>::get_struct_field(ADD_NAME),
+        Option::<Remove>::get_struct_field(REMOVE_NAME),
+        Option::<Metadata>::get_struct_field(METADATA_NAME),
+        Option::<Protocol>::get_struct_field(PROTOCOL_NAME),
+        Option::<SetTransaction>::get_struct_field(SET_TRANSACTION_NAME),
+        Option::<Sidecar>::get_struct_field(SIDECAR_NAME),
+    ])
+    .into()
+});
+
+// Schema of the [`CheckpointMetadata`] action that is included in V2 checkpoints
+// We cannot use `CheckpointMetadata::to_schema()` as it would include the 'tags' field which
+// we're not supporting yet due to the lack of map support.
+static CHECKPOINT_METADATA_ACTION_SCHEMA: LazyLock<SchemaRef> = LazyLock::new(|| {
+    StructType::new([StructField::not_null(
+        CHECKPOINT_METADATA_NAME,
+        DataType::struct_type([StructField::not_null("version", DataType::LONG)]),
+    )])
+    .into()
+});
+
+/// Represents the data needed to create a checkpoint file.
+///
+/// Obtained from [`CheckpointWriter::checkpoint_data`], this struct provides both the
+/// location where the checkpoint file should be written and an iterator over the data
+/// that should be included in the checkpoint.
+///
+/// # Fields
+/// - `path`: The URL where the checkpoint file should be written.
+/// - `data`: A boxed iterator that yields checkpoint actions as chunks of [`EngineData`].
+///
+/// # Usage
+/// 1. Write every action yielded by `data` to persistent storage at the URL specified by `path`.
+/// 2. Ensure that all data is fully persisted before calling `CheckpointWriter::finalize`.
+///    This is crucial to avoid data loss or corruption.
+pub struct CheckpointData {
+    /// The URL where the checkpoint file should be written.
+    pub path: Url,
+
+    /// An iterator over the checkpoint data to be written to the file.
+    pub data: Box<dyn Iterator<Item = DeltaResult<FilteredEngineData>>>,
+}
+
+/// Orchestrates the process of creating and finalizing a checkpoint.
+///
+/// The [`CheckpointWriter`] is the entry point for generating checkpoint data for a Delta table.
+/// It automatically selects the appropriate checkpoint format (V1/V2) based on whether the table
+/// supports the `v2Checkpoints` reader/writer feature.
+///
+/// # Usage Workflow
```

Comment:
seems duplicated from above? consider removing and/or link to the module-level docs?